src/librustc_trans/common.rs

```diff
@@ -434,218 +370,79 @@ impl<'a, 'tcx> FunctionContext<'a, 'tcx> {
-        // Create a fresh builder from the crate context.
-        let llbuilder = unsafe {
-            llvm::LLVMCreateBuilderInContext(ccx.llcx())
-        };
-        OwnedBuilder {
-            builder: Builder {
-                llbuilder: llbuilder,
-                ccx: ccx,
-            }
-        }
-    }
-}
-
-impl<'blk, 'tcx> Drop for OwnedBuilder<'blk, 'tcx> {
-    fn drop(&mut self) {
-        unsafe {
-            llvm::LLVMDisposeBuilder(self.builder.llbuilder);
-        }
-    }
-}
-
-pub struct BlockAndBuilder<'blk, 'tcx: 'blk> {
-    bcx: Block<'blk, 'tcx>,
-    owned_builder: OwnedBuilder<'blk, 'tcx>,
-}
-
-impl<'blk, 'tcx> BlockAndBuilder<'blk, 'tcx> {
-    pub fn new(bcx: Block<'blk, 'tcx>, owned_builder: OwnedBuilder<'blk, 'tcx>) -> Self {
+impl<'a, 'tcx> BlockAndBuilder<'a, 'tcx> {
+    pub fn new(llbb: BasicBlockRef, fcx: &'a FunctionContext<'a, 'tcx>) -> Self {
+        let builder = Builder::with_ccx(fcx.ccx);
         // Set the builder's position to this block's end.
-        owned_builder.builder.position_at_end(bcx.llbb);
+        builder.position_at_end(llbb);
         BlockAndBuilder {
-            bcx: bcx,
-            owned_builder: owned_builder,
+            llbb: llbb,
+            fcx: fcx,
+            builder: builder,
         }
     }
 
-    pub fn with_block<F, R>(&self, f: F) -> R
-        where F: FnOnce(Block<'blk, 'tcx>) -> R
-    {
-        let result = f(self.bcx);
-        self.position_at_end(self.bcx.llbb);
-        result
-    }
-
-    pub fn map_block<F>(self, f: F) -> Self
-        where F: FnOnce(Block<'blk, 'tcx>) -> Block<'blk, 'tcx>
-    {
-        let BlockAndBuilder { bcx, owned_builder } = self;
-        let bcx = f(bcx);
-        BlockAndBuilder::new(bcx, owned_builder)
-    }
-
     pub fn at_start<F, R>(&self, f: F) -> R
-        where F: FnOnce(&BlockAndBuilder<'blk, 'tcx>) -> R
+        where F: FnOnce(&BlockAndBuilder<'a, 'tcx>) -> R
     {
-        self.position_at_start(self.bcx.llbb);
+        self.position_at_start(self.llbb);
         let r = f(self);
-        self.position_at_end(self.bcx.llbb);
+        self.position_at_end(self.llbb);
         r
     }
 
-    // Methods delegated to bcx
-
-    pub fn is_unreachable(&self) -> bool {
-        self.bcx.unreachable.get()
-    }
-
-    pub fn ccx(&self) -> &'blk CrateContext<'blk, 'tcx> {
-        self.bcx.ccx()
+    pub fn ccx(&self) -> &'a CrateContext<'a, 'tcx> {
+        self.fcx.ccx
     }
-    pub fn fcx(&self) -> &'blk FunctionContext<'blk, 'tcx> {
-        self.bcx.fcx()
+    pub fn fcx(&self) -> &'a FunctionContext<'a, 'tcx> {
+        self.fcx
     }
-    pub fn tcx(&self) -> TyCtxt<'blk, 'tcx, 'tcx> {
-        self.bcx.tcx()
+    pub fn tcx(&self) -> TyCtxt<'a, 'tcx, 'tcx> {
+        self.fcx.ccx.tcx()
     }
-    pub fn sess(&self) -> &'blk Session {
-        self.bcx.sess()
+    pub fn sess(&self) -> &'a Session {
+        self.fcx.ccx.sess()
```

Comment:
All of these should probably use `self.ccx` anyway and `fn ccx` above can be removed.